vojtechhabarta/typescript-generator
# Fix #986 (WIP)

# Fix #986

- Adds test in SettingsTest to confirm the issue of ignoring nested generic type arguments. Considering a class such as `List<List<Double>>`, the generic type argument is `List<Double>`, but it was being parsed as List.
- Add SettingsTest.testLoadPrimitiveOrRegularClass
- Adds new tests to CustomTypeMappingTest to confirm the issue of ignoring the generic type argument and mapping the every type that matches the raw class name to the mapped type. For instance, if we configure the plugin to map `List<BigDecimal>` to `number[]`, it was mapping every List attribute to the target type. Even a `List<String>` was being mapped to `number[]` in this example.
- Updates the regex in Settings.parseGenericName. That enables the correct parse of nested generic type arguments. If we have a type such as `Class<T1<T2>, T3>`, type arguments won't be parsed as `T1` and `T3` anymore, but as `T1<T2>` and `T3`. The `Class[T1[T2], T3]` syntax also works.
- Fix the issue of custom type mapping not loading a class that has generic type arguments. The classLoader.loadClass() method requires the raw class name to find the class to load (that is the class name without the generic type arguments). The Settings.loadPrimitiveOrRegularClass now removes the generic parameters from the class name to load the class correctly.
-  Fix the issue in CustomMappingTypeProcessor that was mapping every type that matches the class raw name, ignoring the generic parameters
  ** If we had a mapping `List<BigInteger>:number[]`, every kind of list was being mapped to `number[]`,
  even `List<String>` and any other kind of list.
  ** Update the CustomMappingTypeProcessor.processType() to filter by the generic type arguments
  after checking the raw class name, to confirm the exact type (such as `List<BigInteger>` instead of just `List`)
  before mapping.
  ** The previous change in the default value of Settings.GenericName.typeParameters to an empty list
  makes the filter easier and NPE-safe.
- Update the ModelCompiler.isValidIdentifierPart() to include more valid chars. Since now the class CustomMappingTypeProcessor correctly extracts each generic type argument (even nested ones), a generic type can have some special chars. For instance, in a type such as `java.util.List<java.util.List<String>>`, the generic type argument is `java.util.List<String>`, which has the characters:   `.`    `<`   `>`. Previously, the class was indicating this as an invalid identifier for the type. But this is a valid one, even for Java or TypeScript.
- **Add tests for the new GenericsResolver.typeParameterNameList but I don't know how to make the test work yet.**
